This exercise will require you to pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/ ).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [1]:
# First, import the relevant modules
import requests
import os
API_KEY = os.environ.get('API_PASSWORD')

In [2]:
# Now, call the Quandl API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
url = f'https://www.quandl.com/api/v3/datasets/FSE/AFX_X/data.json?start_date=2017-01-01&end_date=2017-12-31&api_key={API_KEY}'
r = requests.get(url)
json_data = r.json()

In [3]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
type(json_data)

dict

In [4]:
json_data.keys()

dict_keys(['dataset_data'])

In [5]:
json_data['dataset_data'].keys()

dict_keys(['limit', 'transform', 'column_index', 'column_names', 'start_date', 'end_date', 'frequency', 'data', 'collapse', 'order'])

In [6]:
stock_info = json_data['dataset_data']

In [7]:
stock_info['column_names']

['Date',
 'Open',
 'High',
 'Low',
 'Close',
 'Change',
 'Traded Volume',
 'Turnover',
 'Last Price of the Day',
 'Daily Traded Units',
 'Daily Turnover']

In [8]:
keys_to_remove = ['limit', 'transform', 'column_index', 'frequency', 'collapse', 'order', 'start_date', 'end_date']

In [9]:
for key in keys_to_remove:
    stock_info.pop(key)
    

In [10]:
#3. Highest and lowest opening price
def price_list(stock_dict, data_index):
    new_list = []
    for i in range(len(stock_dict['data'])):
        new_list.append(stock_dict['data'][i][data_index])
    return new_list

In [11]:
open_list = price_list(stock_info,1)

In [12]:
open_list[:10]

[51.76, 51.65, 51.45, 51.05, 51.16, 51.88, 52.73, 52.37, 52.7, 53.11]

In [13]:
open_list_rev = [price for price in open_list if price != None]   
max(open_list_rev), min(open_list_rev)

(53.11, 34.0)

Highest opening price was 53.11 dollars and lowest opening pice was 34.00  dollars

In [14]:
#4. What was the largest change in any one day (based on High and Low price)?
high_list = price_list(stock_info, 2)
low_list = price_list(stock_info, 3)
change = [high - low for (high, low) in zip(high_list, low_list)]
change_rev = [delta for delta in change if delta != None]   
max(change_rev)

2.8100000000000023

Largest one-day change was 2.81 dollars

In [15]:
#5. What was the largest change between any two days (based on Closing Price)?
close_list = price_list(stock_info, 4)
close_list[:10]

[51.76, 51.6, 51.82, 51.32, 51.4, 51.27, 51.66, 52.62, 52.01, 52.67]

In [16]:
day_change = [day_2 - day_1 for day_1, day_2 in zip(close_list[:-1], close_list[1:])]
day_change_rev = [delta for delta in day_change if delta != None]   
max(day_change_rev)

2.559999999999995

Largest two-day change was 2.56 dollars

In [17]:
#6. What was the average daily trading volume during this year?
volume_list = price_list(stock_info, 6)
volume_list[:10]

[34640.0,
 40660.0,
 57452.0,
 71165.0,
 120649.0,
 50587.0,
 137313.0,
 129733.0,
 204080.0,
 132981.0]

In [18]:
vol_list_rev = [delta for delta in volume_list if delta != None]   
mean_vol = sum(vol_list_rev)/len(vol_list_rev)
mean_vol


89124.33725490196

Average daily trading volume is 89124

In [19]:
#7. What was the median trading volume during this year
trading_vol = sorted(vol_list_rev)
length = len(trading_vol)
if length % 2 == 0:
    median_vol2 = (trading_vol[length/2 - 1] + trading_vol[length/2]) / 2
else:
    median_vol2 = trading_vol[length//2 - 1]
median_vol2
    
    

74975.0

Median trading volume was 74975

These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)